### ISSUE TITLE
Unaligned memory access on ARM/Unity/IL2CPP

### ISSUE THREAD

**Issue by MiloszKrajewski** [MiloszKrajewski/K4os.Compression.LZ4#40]
Seem like unaligned memory access is still an issue for Unity.
My testing on Raspberry Pi gave me false sense of security as dotnet core runtime (which I used for testing) is handling unaligned access properly while IL2CPP is translating access "as is", and it just ends for Segmentation fault.

The easiest way to handle unaligned access can be done using byte-aligned byte-by-byte access with shifting but it will affect performance a lot. It is easiest thing to do, so to enable people blocked by this issue it will be first approach which I will try to fine-tune later.

> Thanks for reporting this issue. It looks like the code here will do an unaligned memory access, which is not allowed on armv7, hence the crash. Mono works in this case because it generates code that is less efficient than IL2CPP, and does only aligned memory access. With IL2CPP, we have chosen to convert the C# code as-is, so that the generated code will do unaligned access if the C# code does.

**Comment by LittleCodingFox** [comment]
I'll do my best to help you! Let me know when you have a build I can test 😃

**Comment by pepoon** [comment]
xD, yea.. the device it is crashing is samsung s5. I can easily test any build made.

**Comment by MiloszKrajewski** [comment]
Version 1.2.4-alpha which potentially fixes this problem has been release to Nuget/GitHub.

Please, note: to use this safe-aligned version you HAVE TO set `LZ4Codec.EnforceA7` to `true`. 

So, the code of `Main.cs` (which I was using for testing), looked like this:

```
async void Start()
{
	LZ4Codec.EnforceA7 = true;

	Debug.LogWarning($"LZ4Codec.Enforce32: {LZ4Codec.Enforce32}");
	Debug.LogWarning($"LZ4Codec.EnforceA7: {LZ4Codec.EnforceA7}");
	Debug.LogWarning($"Env.Sys64: {Environment.Is64BitOperatingSystem}");
	Debug.LogWarning($"Env.Proc64: {Environment.Is64BitProcess}");
	Debug.LogWarning($"Env.LE: {BitConverter.IsLittleEndian}");

	var bytes = await Load();
	Debug.LogWarning($"Original: {bytes.Length}");
	var pickled = LZ4Pickler.Pickle(bytes);
	Debug.LogWarning($"Pickled: {pickled.Length}");
	var unpickled = LZ4Pickler.Unpickle(pickled);
	Debug.LogWarning($"Unpickled: {unpickled.Length}");
}

public async Task<byte[]> Load()
{
	var decompressedFontPath = Path.Combine(
		Application.streamingAssetsPath, "ab/font_decompress.ab");
	var compressedFontPath = Path.Combine(
		Application.persistentDataPath, "ab/font_compressed.ab");

	var dir = Path.GetDirectoryName(compressedFontPath);
	if (!Directory.Exists(dir)) Directory.CreateDirectory(dir);

	using (var www = new WWW(decompressedFontPath))
	{
		while (!www.isDone) await Task.Delay(100);
		return www.bytes;
	}
}
```

**Comment by pepoon** [comment]
Do you still keep a copy of the project? 
I deleted it from the google drive since i was out of memory after unity team took it.
Would be nice if any of you have the unity project link.

Thanks.

**Comment by MiloszKrajewski** [comment]
[font_decompress.zip](https://github.com/MiloszKrajewski/K4os.Compression.LZ4/files/5048306/font_decompress.zip)
Putting whole project (1GB) is far too much, but the file which was used to test this issue is small enough....

**Comment by pepoon** [comment]
Ok Ill see, Give me time, this month is kind of hell for me. 
Ill try to test it once i get free time.

Thanks Milos.

**Comment by LittleCodingFox** [comment]
I'm currently getting this:
```
Failed to load local remote configuration data file: System.TypeInitializationException: The type initializer for 'K4os.Compression.LZ4.Engine.LL' threw an exception. ---> System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.6.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.
  at K4os.Compression.LZ4.Internal.Mem.Copy (System.Byte* target, System.Byte* source, System.Int32 length) [0x00005] in <ea2f89146db446fbac24b25fb9173539>:0 
  at K4os.Compression.LZ4.Internal.Mem.CloneArray (System.UInt32[] array) [0x0000c] in <ea2f89146db446fbac24b25fb9173539>:0 
  at K4os.Compression.LZ4.Engine.LL..cctor () [0x00038] in <ea2f89146db446fbac24b25fb9173539>:0 
   --- End of inner exception stack trace ---
  at K4os.Compression.LZ4.Encoders.LZ4ChainDecoder..ctor (System.Int32 blockSize, System.Int32 extraBlocks) [0x00060] in <ea2f89146db446fbac24b25fb9173539>:0 
  at K4os.Compression.LZ4.Encoders.LZ4Decoder.CreateChainDecoder (System.Int32 blockSize, System.Int32 extraBlocks) [0x00000] in <ea2f89146db446fbac24b25fb9173539>:0 
  at K4os.Compression.LZ4.Encoders.LZ4Decoder.Create (System.Boolean chaining, System.Int32 blockSize, System.Int32 extraBlocks) [0x00003] in <ea2f89146db446fbac24b25fb9173539>:0 
  at K4os.Compression.LZ4.Streams.LZ4Stream+<>c__DisplayClass2_0.<Decode>b__0 (K4os.Compression.LZ4.Streams.ILZ4Descriptor i) [0x0001d] in <5b2b238d3e9149038c1257b008b15b65>:0 
  at K4os.Compression.LZ4.Streams.LZ4DecoderStream.ReadFrame () [0x0012f] in <5b2b238d3e9149038c1257b008b15b65>:0 
  at K4os.Compression.LZ4.Streams.LZ4DecoderStream.EnsureFrame () [0x00008] in <5b2b238d3e9149038c1257b008b15b65>:0 
  at K4os.Compression.LZ4.Streams.LZ4DecoderStream.Read (System.Byte[] buffer, System.Int32 offset, System.Int32 count) [0x00000] in <5b2b238d3e9149038c1257b008b15b65>:0 
  at System.IO.Stream.InternalCopyTo (System.IO.Stream destination, System.Int32 bufferSize) [0x00012] in <fb001e01371b4adca20013e0ac763896>:0 
  at System.IO.Stream.CopyTo (System.IO.Stream destination) [0x00084] in <fb001e01371b4adca20013e0ac763896>:0 
  at (wrapper remoting-invoke-with-check) System.IO.Stream.CopyTo(System.IO.Stream)
  at FlamingTorchGames.Fairtravel.GameCoordinator.ReloadRemoteConfigurationData () [0x0003c] in C:\GitHub\Fairtravel-Battle\Fairtravel\Assets\Scripts\Game\GameCoordinator.cs:745 
UnityEngine.Debug:LogError(Object)
FlamingTorchGames.Fairtravel.GameCoordinator:ReloadRemoteConfigurationData() (at Assets/Scripts/Game/GameCoordinator.cs:778)
FlamingTorchGames.Fairtravel.<FetchRemoteConfigurationDataCoroutine>d__138:MoveNext() (at Assets/Scripts/Game/GameCoordinator.cs:699)
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
```

**Comment by MiloszKrajewski** [comment]
This looks more like missing assembly: System.Runtime.CompilerServices.Unsafe
Is mono, .net core, il2cpp?

**Comment by LittleCodingFox** [comment]
Mono, Unity 2019.4.8f1

**Comment by LittleCodingFox** [comment]
Does the new code do any sort of codegen? because I wouldn't be able to support iOS that way.

**Comment by MiloszKrajewski** [comment]
No codegen.
I have no idea how Mono/Unity resolves packages, but if it is about drag&drop, then yo need to pull missingl assemblies manually, for example:
https://www.nuget.org/packages/System.Runtime.CompilerServices.Unsafe/

**Comment by pepoon** [comment]
Milos, Is it possible you push a link of 'Assets' folder of the old unity project? See if it is small in size.
Thanks.

**Comment by MiloszKrajewski** [comment]
"It works on my machine". Both using IL2CPP and Mono. All assemblies (dependencies) zipped.
[Plugins.zip](https://github.com/MiloszKrajewski/K4os.Compression.LZ4/files/5081265/Plugins.zip)

**Comment by MiloszKrajewski** [comment]
Whole project will be available for few days here: https://tinyurl.com/yxarbkwc

**Comment by LittleCodingFox** [comment]
Just wanted to give you a heads up that I've been extremely busy lately and that's why there's no news. I'll do my best to test this ASAP, but I don't have an ETA yet.

Thank you for your patience 🙏

**Comment by pepoon** [comment]
yea me too, im running out of deadline.
i have the files on my pc,  ill test  it as well once things calm down

**Comment by MiloszKrajewski** [comment]
Addressed in 1.2.5
